[IIN][FBMN] Method to collapse ion identity networks to single nodes

robinschmid opened the pull request:
Hey everybody,

this change adds a new argument to convert_networks_to_graphml.py

We can run it twice
1. generating the standard graphml
2. generating a graphml after collapsing all ion identity networks into single nodes

**IIN collapsing**
1. find all ion identity networks by ID
2. merge all nodes into the node with the highest node intensity
3. redirect all edges to this node, remove the other nodes from the network
4. add more columns (node attributes) to reflect the "summed ion intensity", intensity of each ion adduct, number of merged nodes

I still have to finish the unit test. How can I run the method/test to see if the graphml is created?

Best regards 
Robin

mwang87 commented:
Would it make more sense to make this generic, so we could collapse based upon an arbitrary feature?

mwang87 commented on feature-based-molecular-networking/test/test_workflow_components.py:
```diff
@@ -39,6 +39,21 @@ def test_additional_edges():
+                                               "iin.graphml", True)
+
+    G = nx.read_graphml("iin.graphml")
+    # TODO: add test values
```
Assert the network size is smaller than before

robinschmid replied:
For now I just took the number of rows from the cluster_summary.tsv - this should be the initial number of nodes, right?

robinschmid commented on shared_code/molecular_network_filtering_library.py:
```diff
@@ -83,7 +83,9 @@ def loading_network(filename, hasHeaders=False):
         intermediate_graph_nodes.add(edge_object["node2"])
 
-        intermediate_edges_to_add.append((edge_object["node1"], edge_object["node2"], edge_object))
+        # set key to edgetype for options to remove edge later in case multiple edges are present
```
Is needed to remove edges from a MultiGraph

robinschmid commented on shared_code/molecular_network_filtering_library.py:
```diff
@@ -117,7 +119,10 @@ def add_additional_edges(G, path_to_supplemental_edges):
-            edges_to_add.append((node1, node2, edge_object))
+
+            #set key to edgetype for options to remove edge later
+            edge_key = edgetype
```
The key is needed to remove edges from a MultiGraph

robinschmid commented on feature-based-molecular-networking/tools/feature-based-molecular-networking/scripts/convert_networks_to_graphml.py:
```diff
@@ -5,39 +5,60 @@
+    # add "neutral molecule" nodes and collapse IIN edges
+    # IIN edges = supplemental pairs (input_pairsfolder)
     parser.add_argument('--input_pairsfolder', help='input_pairsfolder')
+    parser.add_argument('collapse_additional_edges', help='collapse_additional_edges')
```
Made optional by --collapse_additional_edges and added default="False"

robinschmid commented on feature-based-molecular-networking/tools/feature-based-molecular-networking/scripts/convert_networks_to_graphml.py:
```diff
@@ -5,39 +5,60 @@
 
-    create_graphml(args.input_pairs, args.input_clusterinfosummary, args.input_librarysearch, args.input_analoglibrarysearch, args.input_pairsfolder, args.output_graphml)
+    # boolean
+    collapse_additional_edges = (args.collapse_additional_edges is not None)
```
Made optional by --collapse_additional_edges and added default="False"

robinschmid commented on .github/workflows/workflow-unit-fbmn.yml:
```diff
@@ -20,4 +20,4 @@ jobs:
       run: |
         pip install nose2
-        cd feature-based-molecular-networking/test && nose2 -v
+        cd feature-based-molecular-networking/test && nose2 -v -c unittest.cfg
```
Added config file for logging in tests

robinschmid commented on feature-based-molecular-networking/tools/feature-based-molecular-networking/scripts/constants.py:
```diff
@@ -0,0 +1,42 @@
+# constants for....
+
+# edges
+class EDGE:
+    TYPE_ATTRIBUTE = "EdgeType"
+    ION_TYPE = "MS1 annotation"  # ion identity molecular networking IIMN
+    COSINE_TYPE = "Cosine"  # molecular networking edge
+    # list of all types
+    TYPES = [ION_TYPE, COSINE_TYPE]
+
+    SCORE_ATTRIBUTE = "EdgeScore"  # score for all edge types (cosine, ion identity, ..)
+    NODE1_ATTRIBUTE = "node1"  #
+    NODE2_ATTRIBUTE = "node2"  #
+    MASS_DIFF_ATTRIBUTE = "mass_difference"  #
+    ANNOTATION_ATTRIBUTE = "EdgeAnnotation"  # edge annotation for all edge types
+    EXPLAINED_INTENSITY_ATTRIBUTE = "explained_intensity"  # explained intensity between the two MS2 spectra from MN
```
Added constants.py to store and access node and edge attributes and values
import constants as CONST
CONST.EDGE.TYPE_ATTRIBUTE
CONST.EDGE.SCORE_ATTRIBUTE
CONST.NODE. ...

robinschmid commented on feature-based-molecular-networking/tools/feature-based-molecular-networking/scripts/ion_network_utils.py:
```diff
@@ -0,0 +1,206 @@
+import constants as CONST
+import operator
+
+logger = logging_utils.get_logger(__name__)
```
Easy creation of logger

robinschmid commented:
## Logging
- added logging_utils.py for easy creation of loggers `logger = logging_utils.get_logger(__name__)`
- added -c unittest.cfg to the nose2 config for logging in tests
- **TODO: Define output log file in logging_utils.py**

## constants.py
- added to hold "constants"/references to node/edge attribute strings. **Examples**:
  - constants.EDGE.SCORE_ATTRIBUTE (dict key to access score)
  - constants.NODE.ABUNDANCE_ATTRIBUTE (dict key to access "sum(precursor intensity)")
- **TODO: Might want to slowly adapt the constants.py in other parts of GNPS (e.g., network generation)**

## Collapsing sub networks (e.g., ion identity networks) based on node attributes
- modularized the methods in ion_network_utils.py more
- `ion_network_utils.collapse_based_on_node_attribute(H, merge_att, best_node_att, best_edge_att=CONST.EDGE.SCORE_ATTRIBUTE,edge_comparator=operator.ge):`
  - define node attribute to merge all nodes with the same value (e.g., ion network ID)
  - define how to merge edges (keep edge with highest score)

robinschmid commented:
### TODO
- move logging_utils.py, ion_network_utils.py, constants.py to **shared_code**
- create symlink (_my symlinks did not work_)
- does it also work if we use one of the two proposed package linking options from ... to link shared_code:
    https://stackoverflow.com/questions/448271/what-is-init-py-for

robinschmid commented:
Still need to add to the workflow:

    convert_networks_to_graphml.create_graphml("reference_data/IIN/edges.tsv",
                                               "reference_data/IIN/cluster_summary.tsv",
                                               "reference_data/IIN/library_matches.tsv",
                                               "reference_data/IIN/library_matches.tsv",
                                               "reference_data/IIN/additional_edges",
                                               "iin.graphml", True)

robinschmid commented:
Test job
https://gnps.ucsd.edu/ProteoSAFe/status.jsp?task=b22b6c3cc0da451ba28f2912547168fd
